File: python/langchain_kubernetes/manifests.py
"""Kubernetes manifest builders for sandbox Pods, Namespaces, and NetworkPolicies."""

from __future__ import annotations

from typing import Any

from langchain_kubernetes._utils import common_labels
from langchain_kubernetes.config import KubernetesProviderConfig


def build_namespace_manifest(name: str, sandbox_id: str) -> dict[str, Any]:
    """Return a Namespace manifest dict.

    Args:
        name: Namespace name.
        sandbox_id: Sandbox identifier to embed in labels.

    Returns:
        Kubernetes Namespace manifest as a plain Python dict.
    """
    return {
        "apiVersion": "v1",
        "kind": "Namespace",
        "metadata": {
            "name": name,
            "labels": common_labels(sandbox_id),
        },
    }


def build_pod_manifest(
    *,
    pod_name: str,
    namespace: str,
    sandbox_id: str,
    config: KubernetesProviderConfig,
) -> dict[str, Any]:
    """Return a Pod manifest dict that runs ``sleep infinity``.

    The Pod is configured with secure defaults: non-root user, all Linux
    capabilities dropped, no privilege escalation, RuntimeDefault seccomp
    profile, and no automounted service-account token.

    Args:
        pod_name: Name for the Pod resource.
        namespace: Namespace in which to create the Pod.
        sandbox_id: Sandbox identifier embedded in labels.
        config: Provider configuration driving image, resources, env vars, etc.

    Returns:
        Kubernetes Pod manifest as a plain Python dict.
    """
    labels = common_labels(sandbox_id)

    # Build resource requirements (omit keys whose values are None)
    resources: dict[str, Any] = {}
    requests: dict[str, str] = {}
    limits: dict[str, str] = {}
    if config.cpu_request:
        requests["cpu"] = config.cpu_request
    if config.memory_request:
        requests["memory"] = config.memory_request
    if config.cpu_limit:
        limits["cpu"] = config.cpu_limit
    if config.memory_limit:
        limits["memory"] = config.memory_limit
    if requests:
        resources["requests"] = requests
    if limits:
        resources["limits"] = limits

    # Build environment variables
    env = [{"name": k, "value": v} for k, v in config.extra_env.items()]

    security_context: dict[str, Any] = {
        "allowPrivilegeEscalation": False,
        "runAsNonRoot": True,
        "capabilities": {"drop": ["ALL"]},
    }
    if config.run_as_user is not None:
        security_context["runAsUser"] = config.run_as_user
    if config.run_as_group is not None:
        security_context["runAsGroup"] = config.run_as_group
    if config.seccomp_profile is not None:
        security_context["seccompProfile"] = {"type": config.seccomp_profile}

    container: dict[str, Any] = {
        "name": config.container_name,
        "image": config.image,
        "imagePullPolicy": config.image_pull_policy,
        "command": ["sleep", "infinity"],
        "securityContext": security_context,
    }
    if resources:
        container["resources"] = resources
    if env:
        container["env"] = env

    return {
        "apiVersion": "v1",
        "kind": "Pod",
        "metadata": {
            "name": pod_name,
            "namespace": namespace,
            "labels": labels,
        },
        "spec": {
            "restartPolicy": "Never",
            "automountServiceAccountToken": False,
            "containers": [container],
        },
    }


def build_network_policy_manifest(
    *,
    namespace: str,
    sandbox_id: str,
) -> dict[str, Any]:
    """Return a deny-all NetworkPolicy manifest for a sandbox Pod.

    The policy matches Pods by the ``deepagents.langchain.com/sandbox-id``
    label and blocks all ingress and egress traffic.

    Args:
        namespace: Namespace where the NetworkPolicy will live.
        sandbox_id: Sandbox identifier used for both the name suffix and
            the pod-selector label.

    Returns:
        Kubernetes NetworkPolicy manifest as a plain Python dict.
    """
    from langchain_kubernetes._utils import LABEL_SANDBOX_ID

    return {
        "apiVersion": "networking.k8s.io/v1",
        "kind": "NetworkPolicy",
        "metadata": {
            "name": f"deepagents-sandbox-deny-all-{sandbox_id}",
            "namespace": namespace,
            "labels": common_labels(sandbox_id),
        },
        "spec": {
            "podSelector": {
                "matchLabels": {LABEL_SANDBOX_ID: sandbox_id},
            },
            "policyTypes": ["Ingress", "Egress"],
            "ingress": [],
            "egress": [],
        },
    }

File: typescript/src/manifests.ts
import * as k8s from "@kubernetes/client-node";
import type { KubernetesProviderConfig } from "./config.js";
import { sandboxLabels, networkPolicyName } from "./utils.js";

/**
 * Build a Pod manifest for a new sandbox.
 *
 * @param podName    - The Pod name (also used as part of the sandbox ID).
 * @param namespace  - The namespace the Pod will live in.
 * @param sandboxId  - The logical sandbox ID (used for labels).
 * @param config     - Provider configuration controlling image, resources, etc.
 */
export function buildPodManifest(
  podName: string,
  namespace: string,
  sandboxId: string,
  config: KubernetesProviderConfig
): k8s.V1Pod {
  const labels = sandboxLabels(sandboxId);

  const envVars: k8s.V1EnvVar[] = config.env
    ? Object.entries(config.env).map(([name, value]) => ({ name, value }))
    : [];

  const container: k8s.V1Container = {
    name: "sandbox",
    image: config.image,
    imagePullPolicy: config.imagePullPolicy,
    command: config.command,
    workingDir: config.workdir,
    env: envVars.length > 0 ? envVars : undefined,
    resources: {
      requests: {
        cpu: config.cpuRequest,
        memory: config.memoryRequest,
      },
      limits: {
        cpu: config.cpuLimit,
        memory: config.memoryLimit,
        "ephemeral-storage": config.ephemeralStorageLimit,
      },
    },
    securityContext: {
      runAsNonRoot: true,
      runAsUser: config.runAsUser,
      runAsGroup: config.runAsGroup,
      allowPrivilegeEscalation: false,
      capabilities: {
        drop: ["ALL"],
      },
      seccompProfile: {
        type: config.seccompProfile as k8s.V1SeccompProfile["type"],
      },
    },
    volumeMounts: config.volumeMounts,
  };

  const imagePullSecrets: k8s.V1LocalObjectReference[] | undefined =
    config.imagePullSecrets?.map((name) => ({ name }));

  const podSpec: k8s.V1PodSpec = {
    containers: [container],
    restartPolicy: "Never",
    automountServiceAccountToken: false,
    serviceAccountName: config.serviceAccount,
    imagePullSecrets,
    nodeSelector: config.nodeSelector,
    tolerations: config.tolerations,
    volumes: config.volumes,
    initContainers: config.initContainers,
    ...config.podTemplateOverrides,
  };

  return {
    apiVersion: "v1",
    kind: "Pod",
    metadata: {
      name: podName,
      namespace,
      labels,
      annotations: config.podTtlSeconds
        ? { "deepagents.langchain.com/ttl-seconds": String(config.podTtlSeconds) }
        : undefined,
    },
    spec: podSpec,
  };
}

/**
 * Build a Namespace manifest for the per-sandbox namespace strategy.
 *
 * @param namespaceName - The namespace name to create.
 * @param sandboxId     - The logical sandbox ID (used for labels).
 * @param config        - Provider configuration.
 */
export function buildNamespaceManifest(
  namespaceName: string,
  sandboxId: string,
  config: KubernetesProviderConfig
): k8s.V1Namespace {
  const labels: Record<string, string> = {
    ...sandboxLabels(sandboxId),
    ...config.namespaceLabels,
  };

  return {
    apiVersion: "v1",
    kind: "Namespace",
    metadata: {
      name: namespaceName,
      labels,
    },
  };
}

/**
 * Build a deny-all NetworkPolicy manifest that blocks all ingress and egress
 * for the sandbox Pod identified by `sandboxId`.
 *
 * @param namespace - The namespace the policy lives in.
 * @param sandboxId - The logical sandbox ID, used in label selectors and the policy name.
 */
export function buildNetworkPolicyManifest(
  namespace: string,
  sandboxId: string
): k8s.V1NetworkPolicy {
  return {
    apiVersion: "networking.k8s.io/v1",
    kind: "NetworkPolicy",
    metadata: {
      name: networkPolicyName(sandboxId),
      namespace,
      labels: sandboxLabels(sandboxId),
    },
    spec: {
      podSelector: {
        matchLabels: {
          "deepagents.langchain.com/sandbox-id": sandboxId,
        },
      },
      policyTypes: ["Ingress", "Egress"],
      ingress: [],
      egress: [],
    },
  };
}

/**
 * Build a ResourceQuota manifest that caps aggregate resource usage within
 * a namespace (used for the per-sandbox namespace strategy).
 *
 * @param namespace - The namespace the quota applies to.
 * @param sandboxId - The logical sandbox ID, used for labels.
 * @param config    - Provider configuration for resource limits.
 */
export function buildResourceQuotaManifest(
  namespace: string,
  sandboxId: string,
  config: KubernetesProviderConfig
): k8s.V1ResourceQuota {
  return {
    apiVersion: "v1",
    kind: "ResourceQuota",
    metadata: {
      name: "deepagents-sandbox-quota",
      namespace,
      labels: sandboxLabels(sandboxId),
    },
    spec: {
      hard: {
        pods: "1",
        "requests.cpu": config.cpuRequest,
        "requests.memory": config.memoryRequest,
        "limits.cpu": config.cpuLimit,
        "limits.memory": config.memoryLimit,
        "requests.ephemeral-storage": config.ephemeralStorageLimit,
      },
    },
  };
}
